src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -236,24 +160,174 @@ void addArg(RefKind refKind, BoundExpression expression)
                             _localRewriter._diagnostics.Add(t.Syntax, useSiteDiagnostics);
                             if (conversion.Exists)
                             {
-                                _sideEffectBuilder.Add(_factory.AssignmentExpression(output, _factory.Convert(type, input, conversion)));
+                                if (conversion.Kind == ConversionKind.ExplicitNullable &&
+                                    inputType.GetNullableUnderlyingType().Equals(output.Type, TypeCompareKind.AllIgnoreOptions) &&
+                                    _localRewriter.TryGetNullableMethod(t.Syntax, inputType, SpecialMember.System_Nullable_T_GetValueOrDefault, out MethodSymbol getValueOrDefault))
+                                {
+                                    // As a special case, since the null test has already been done we can use Nullable<T>.GetValueOrDefault
+                                    return _factory.AssignmentExpression(output, _factory.Call(input, getValueOrDefault));
+                                }
+                                else
+                                {
+                                    return _factory.AssignmentExpression(output, _factory.Convert(type, input, conversion));
+                                }
                             }
                             else
                             {
-                                _sideEffectBuilder.Add(_factory.AssignmentExpression(output, _factory.As(input, type)));
+                                return _factory.AssignmentExpression(output, _factory.As(input, type));
```

Comment:
>return _factory.AssignmentExpression(output, [](start = 32, length = 44)

Consider moving this outside the `if`s and calculate the RHS expression only within the `if`s.